File: lib/setvar.ex
# fixed size set
defmodule Setvar do
  require Record

  Record.defrecord(:setvar,
    is_fixed: false,
    size: 10000,
    domain: IntegerDomain.new(0, 10000),
    values: %MapSet{}
  )

  @type setvar_int() :: record(:setvar, is_fixed: boolean(), size: integer(), values: %MapSet{})
  @type setvar() :: {Setvar, setvar_int()}
  @behaviour Var
  def new_intset(size, mindomain, maxdomain) do
    {Setvar,
     setvar(
       size: size,
       domain: IntegerDomain.new(mindomain, maxdomain)
     )}
  end

  # removes value from possible values for the set
  def forbid(value, {Setvar, a = setvar(domain: d)}) do
    d = IntegerDomain.forbid(value, d)
    {Setvar, setvar(a, domain: d)}
  end

  def require(value, {Setvar, a = setvar(size: size, domain: d, is_fixed: fixed, values: values)}) do
    case MapSet.member?(values, value) do
      true ->
        {Setvar, a}

      false ->
        case fixed do
          true ->
            :failed

          false ->
            case IntegerDomain.possible(value, d) do
              false ->
                :failed

              true ->
                {Setvar,
                 setvar(a,
                   values: MapSet.put(values, value),
                   is_fixed: size == MapSet.size(values) + 1
                 )}
            end
        end
    end
  end

  def split({Setvar, setvar(is_fixed: true)}) do
    :failed
  end

  def split({Setvar, var = setvar(values: v, size: size, domain: domain)}) do
    min = IntegerDomain.min(domain)
    max = IntegerDomain.max(domain)
    new = find(min, max, v)
    fixed = size == MapSet.size(v) + 1

    {{Setvar, setvar(var, values: MapSet.put(v, new), is_fixed: fixed)},
     {Setvar, setvar(var, values: v, domain: IntegerDomain.forbid(new, domain))}}
  end

  defp find(curr, max, v) do
    case MapSet.member?(v, curr) do
      false -> curr
      true -> find(curr + 1, max, v)
    end
  end

  def required({Setvar, setvar(values: v)}) do
    v
  end

  def possible(value, {Setvar, setvar(domain: d)}) do
    IntegerDomain.possible(value, d)
  end

  @spec unify(setvar(), setvar()) :: :failed | setvar()
  def unify(a = {Setvar, v}, {Setvar, v}), do: a

  def unify(
        {Setvar, a = setvar(domain: d1, size: s, values: v1)},
        {Setvar, setvar(domain: d2, size: s, values: v2)}
      ) do
    v = MapSet.union(v1, v2)

    case {MapSet.size(v), s} do
      {a, b} when a > b ->
        :failed

      {equal, equal} ->
        domain = IntegerDomain.unify(d1, d2)
        {Setvar, setvar(a, domain: domain, is_fixed: true, values: v)}

      {_, _} ->
        {Setvar, setvar(a, domain: IntegerDomain.unify(d1, d2), values: v)}
    end
  end

  def unify(_, _), do: :failed

  def is_fixed({Setvar, setvar(is_fixed: v)}), do: v
  def value_if_fixed({Setvar, v = setvar(is_fixed: true)}), do: v
  def value_if_fixed({Setvar, setvar(is_fixed: false)}), do: :undefined
end

File: src/setvar.erl
-module(setvar).

-export([new/3, forbid/2, require/2, split/1, possible/2, unify/2,
         is_fixed/1, value_if_fixed/1, disjoint/2]).

-behaviour('Elixir.Var').

-record(setvar,
        {is_fixed = false  :: boolean(), size = 10000  :: integer(),
         domain = 'Elixir.IntegerDomain':new(0, 10000)  ::
           'Elixir.IntegerDomain':intdomain(),
         values = sets:new()  :: sets:set()}).

new(Size, Min, Max) ->
  #setvar{size = Size, domain = 'Elixir.IntegerDomain':new(Min, Max)}.

forbid(N, V = #setvar{domain = D}) ->
  D1 = 'Elixir.IntegerDomain':forbid(N, D), V#setvar{domain = D1}.

require(N, Var = #setvar{size = S, domain = D, is_fixed = F, values = V}) ->
  case sets:is_element(N, V) of
    true -> Var;
    false ->
      case F of
        true -> failed;
        false ->
          case 'Elixir.IntegerDomain':possible(N, D) of
            false -> failed;
            true -> Var#setvar{is_fixed = S == sets:size(V) + 1, values = sets:add_element(N, V)}
          end
      end
  end.

disjoint(#setvar{ values = V1}, #setvar{ values = V2}) ->
  sets:is_disjoint(V1, V2).

split(#setvar{is_fixed = true}) -> failed;
split(Var = #setvar{size = S, domain = D, values = V}) ->
  Min = 'Elixir.IntegerDomain':min(D),
  Max = 'Elixir.IntegerDomain':max(D),
  case find(Min, Max, V) of
    failed -> failed;
    New ->
      Fixed = S == sets:size(V) + 1,
      {Var#setvar{values = sets:add_element(New, V), is_fixed = Fixed},
       Var#setvar{domain = 'Elixir.IntegerDomain':forbid(New, D)}}
  end.

find(C, C, _) -> failed;
find(C, M, V) ->
  case sets:is_element(C, V) of
    false -> C;
    true -> find(C + 1, M, V)
  end.

possible(V, #setvar{domain = D}) -> 'Elixir.IntegerDomain':possible(V, D).

unify(A, A) -> A;
unify(A = #setvar{domain = D1, size = S, values = V2},
      #setvar{domain = D2, size = S, values = V1}) ->
  V = sets:union(V1, V2),
  case sets:size(V) of
    VS when VS > S -> failed;
    S2 ->
      case 'Elixir.IntegerDomain':unify(D1, D2) of
        {} -> failed;
        D -> A#setvar{domain = D, is_fixed = S2 == S, values = V}
      end
  end;
unify(_, _) -> failed.

is_fixed(#setvar{is_fixed = F}) -> F.

value_if_fixed(#setvar{is_fixed = false}) -> undefined;
value_if_fixed(#setvar{values = V}) -> sets:to_list(V).

